### ISSUE TITLE
Uncaught ETIMEDOUT exceptions

### ISSUE THREAD

**Issue by SukantGujar** [pouchdb/pouchdb#5114]
During a local Pouch DB (5.3.2) from remote DB (CouchDB 1.6.1) replication call, I am observing repeated exceptions -
![capture](https://cloud.githubusercontent.com/assets/6153264/14940310/2158a7a8-0f8f-11e6-8242-2500f104052f.PNG)

The live replication call code is bare minimum -

`localDB.replicate.from(remoteURL, { live: true, retry:true })`

Is this expected? If so, could PochDB handle the long polling timeouts in a manner which doesn't end up with uncaught exceptions?

**Comment by nolanlawson** [comment]
Retry replication is supposed to gracefully handle timeouts. If it's throwing an exception and halting replication, then that's a bug.

**Comment by SukantGujar** [comment]
The exception is reported by the browser as uncaught and then bubbles up as err in the paused event.
On May 2, 2016 6:27 AM, Nolan Lawson <notifications@github.com> wrote:Retry replication is supposed to gracefully handle timeouts. If it's throwing an exception and halting replication, then that's a bug.

**Comment by nolanlawson** [comment]
Can you provide a test case to reproduce?

**Comment by SukantGujar** [comment]
I am not doing anything specific. Its a plain CouchDB setup in Windows 10 as the backend. From PouchDB client in Chrome, I am trying to replicate a remoteDB into a local instance. As you will notice in the screenshot, the exception is reporting after a _changes longpoll call is waited on till it times out. If you need, I can provide the Fiddler trace too.

**Comment by nolanlawson** [comment]
A test case would still be very helpful. I've never seen this error before, so it might be something related to CouchDB's Windows 10 build.

**Comment by SukantGujar** [comment]
Sure. I am assuming you are referring to a JS test case. Which test framework is preferable? I can try to rig something up in Jasmine. However, if this turns out to be reproduce-able on specific machines only, I am willing to share a screen recording with logs as well. If needed, I can do a live demo too.
Thanks.
 

    On Tuesday, 3 May 2016 9:20 PM, Nolan Lawson <notifications@github.com> wrote:
 
 

 A test case would still be very helpful. I've never seen this error before, so it might be something related to CouchDB's Windows 10 build.—

**Comment by nolanlawson** [comment]
A live web site would probably be best, but I can understand if that's hard to set up. Otherwise instructions for a Windows VM or something would work too. Else if you can fix the error yourself and open a PR that would be even better. :D

**Comment by SukantGujar** [comment]
I will try to have my installation available on a public IP. 

    On Tuesday, 3 May 2016 11:46 PM, Nolan Lawson <notifications@github.com> wrote:
 
 

 A live web site would probably be best, but I can understand if that's hard to set up. Otherwise instructions for a Windows VM or something would work too. Else if you can fix the error yourself and open a PR that would be even better. :D—

**Comment by easyservices** [comment]
Hi there. I have the same issue for the same setup and use case. My apps is *simply* trying to sync-up CouchDB and PouchDB in bidirectional mode. It does not happen on Win7 (Chrome 40.x) but it does happen on Win10 (I tried under Chrome 50.xx & IE/Edge 11.x => same issue). I have CouchDB build 1.6.1 and PouchDB build 5.3.2. Surprisingly it does happen randomly or to be more precise: systematically when you update within a short timeframe the same document.
Error message :
pouchdb-5.3.2.js:5512 Uncaught (in promise) Error: ETIMEDOUT(…)xhr.onreadystatechange @ pouchdb-5.3.2.js:5512timeoutReq @ pouchdb-5.3.2.js:5431
pouchdbDAO.js:23 paused

**Comment by daleharvey** [comment]
I have never seen this on Mac, it is sounding like a win10 specific issue, I am quite surprised that the OS is affecting this though. 

As for ETIMEDOUT, the `_changes` request is a long poll request that times out after 25 second however receives a heartbeat every 10 seconds so it should never time out at all, if ie and chrome use the same networking stack under the hood on windows then it sounds like that networking stack handles timeouts + heartbeats in a different way, throwing an uncaught ETIMEDOUT whatever we do. This should not affect the reliability of replication, a timed out request will trigger a new request, but if anyone with access to win10 can figure out how to prevent the exception being shown it would be helpful for users

**Comment by easyservices** [comment]
Hmmm a correction to my previous post: it **does** happen as well on Windows 7 with Chrome 50.xx and Edge/IE11 but I can confirm it does not happen on Windows XP. So it does happen on Windows 7 and 10 (x64) but not on Windows XP. If it can help...

**Comment by dharders** [comment]
I've experienced this bug with **pouchdb-server** as my remote.

But not CouchDB 1.6.1. The only way I could reproduce this bug on CouchDB 1.6.1 was by setting `{ ..., timeout: 7000 }` (a timeout value less than heartbeat default of 10000). 

Note that the value you set for `timeout` is **ALWAYS 5000ms less** on the XHR _changes GET request. I'm not sure if this is a bug, or normal behaviour.

Eg for the above timeout of 7000 here is pouchdb's _changes feed req:

`"http://192.168.1.180:5984/testdb/_changes?timeout=2000&style=all_docs&feed=longpoll&heartbeat=10000&since=3239&limit=10000"`

Note the timeout of 2000, not 7000. Is this expected behaviour @daleharvey ? 

It's always 5000 less than the specified timeout. i.e. timeout: 50000 -> _changes?timeout=45000...

**TLDR:** I can confirm that this ETIMEDOUT error does bubble up to the `paused` event. I've only really found it occur with a pouchdb-server remote (not CouchDB, except trying with an intentionally low timeout value).

Note: I use the `sync()` API with [this unofficial fix #4592](https://github.com/pouchdb/pouchdb/issues/4592) to propagate errors to the `paused` event as currently it doesn't for `sync()`, although it says it does in the [docs](https://pouchdb.com/api.html#sync)

My Setup: Win7x64, PouchDB 5.4.0, CouchDB 1.6.1, pouchdb-server 1.1.1

**Comment by dharders** [comment]
OK 5000ms timeout difference is expected behaviour: [pouchdb-adapter-http/src/index.js#L734-L736](https://github.com/pouchdb/pouchdb/blob/master/packages/pouchdb-adapter-http/src/index.js#L734-L736). Sorry for the confusion.

Getting back on topic:

Should the ETIMEDOUT error ever legitimately reach the paused event, since a retry will occur straight away anyway?

I use the pause with error event, i.e. `pause(err){ if (err)offline = true}`, as my 'went offline' trigger. So currently I get a WENT OFFLINE flash every X (timeout) milliseconds.

**Comment by yaronyg** [comment]
When I ran Express-PouchDB tests on my Mac (e.g. no browser anywhere) I got this as well. Please see https://github.com/pouchdb/express-pouchdb/issues/328 which provides the repo.

**Comment by yaronyg** [comment]
I believe the line at fault is [here](https://github.com/pouchdb/pouchdb/blob/ebbadb7dd7c88c032cc727f405276ba8cc60f48d/packages/pouchdb-core/src/changes.js#L178). What appears to be happening is that we get the timeout error (or any error actually) on the info() call and then that line of code just throws. Up the stream the throw will be turned into a reject() by the wrapping promise so this means that in effect the catch handler in the previous line has itself returned a reject and there is nobody listening for it and so... boom... un-handled rejection.

Based on what I see us doing [here](https://github.com/pouchdb/pouchdb/blob/ebbadb7dd7c88c032cc727f405276ba8cc60f48d/packages/pouchdb-core/src/changes.js#L165), I think the fix is to change the line to `callback(err)` instead of a throw.

If this sounds reasonable I can submit a PR but I have a question about how to test it. The obvious thing to do is to try to replicate with a non-existent URL and make sure we get an error event and no unhandled promise rejections.

To Summarize:
Question 1 - Do y'all agree that the bug is caused by [this line?](https://github.com/pouchdb/pouchdb/blob/ebbadb7dd7c88c032cc727f405276ba8cc60f48d/packages/pouchdb-core/src/changes.js#L178)
Question 2 - If so, do you agree that the fix is to replace that line with `callback(err)`?
Question 3 - To test this do you think trying to replicate with a non-existent URL and making sure we get an error event and no un-handled rejections is acceptable?

**Comment by nolanlawson** [comment]
1. That line, since it is ignored by istanbul, is extremely suspicious
2. Dunno
3. Yes, sounds reasonable. Let's test it.

**Comment by provedotit** [comment]
I've seen users trigger this on both Mac and Windows 7/10 - Chrome, Firefox, IE and Safari.

**Comment by jefbarn** [comment]
Seeing this when I tried deploying CouchDB to Windows Server 2012 today.
Host is Chrome Windows 7.

**Comment by daleharvey** [comment]
> Question 3 - To test this do you think trying to replicate with a non-existent URL and 
> making sure we get an error event and no un-handled rejections is acceptable?

We already have this test @ https://github.com/pouchdb/pouchdb/blob/master/tests/integration/test.replication.js#L4445 so would need to clarify what is different from that and be specific about the case in the test

**Comment by ssured** [comment]
I do see these errors coming into my rollbar logger. Nothing helpful in there yet tbh. Hope to discover a pattern and will then update here.

**Comment by yaronyg** [comment]
@daleharvey To catch the error caused [here](https://github.com/pouchdb/pouchdb/blob/ebbadb7dd7c88c032cc727f405276ba8cc60f48d/packages/pouchdb-core/src/changes.js#L178) one would have to be doing continuous replication with a since value other than 'now' where the HTTP request to get info fails. None of that applies to [this test](https://github.com/pouchdb/pouchdb/blob/master/tests/integration/test.replication.js#L4445). The server it is testing against exists so info won't fail. And the replication is not continuous. I am creating a test that uses a bogus URL that will fail info and thus test the problem.

And as punishment for trying to fix this my test now throws up a new uncaught ECONNREFUSED error that appears to come from [here](https://github.com/pouchdb/pouchdb/blob/0d384e848f0ec5238c2f816393b5535ae2bb0d6e/packages/pouchdb-checkpointer/src/index.js#L107). And yes, I do get the hint that I need to look for 'throw err' everywhere.

**Comment by yaronyg** [comment]
@daleharvey My new plan is to just write a test that directly calls Changes.prototype.doChanges and shows that things behave themselves.

**Comment by daleharvey** [comment]
@yaronyg Did https://github.com/pouchdb/pouchdb/commit/b191d88448a6c54cfee976c770da5315ab7d0709 fix this? or was it something additional you found while investigating

**Comment by yaronyg** [comment]
@daleharvey It was something additional I found while trying to fix this one. I should get to this one today.

**Comment by yaronyg** [comment]
@daleharvey & @nolanlawson I'm sorry to bug you guys but I've run into something that has me scratching my head. 

I'm specifically looking at [this](https://github.com/pouchdb/pouchdb/blob/67c9c5c89bf398dd0259e4d5c6ccf663e56a45b3/packages/pouchdb-core/src/changes.js#L171) section of code. First off it's weird because it's doing something async that sets a value that one would naively think would control how the changes code works but nothing is held up waiting for the response. So clearly whatever it's doing can't be that important. 

But when I went looking for who uses startSeq all I can find is the section of code [here](https://github.com/pouchdb/pouchdb/blob/67c9c5c89bf398dd0259e4d5c6ccf663e56a45b3/packages†/pouchdb-core/src/changes.js#L72) which looks like a NOOP.

So it seems to me that both of the sections of code I reference above aren't needed. I disabled both sections of code and ran all the tests and everything passed. So my current plan is to just submit a PR deleting those two sections of code.

**Comment by yaronyg** [comment]
So #5457 resolved the version of this problem I have been running into. And looking at the stack that @SukantGujar posted, which includes doChanges, I wouldn't be surprised if it doesn't fix the issue there as well. But since we never got an official repo from @SukantGujar I'm not sure what the process is for closing this bug.

**Comment by daleharvey** [comment]
It looks like its fixed to me, closing out, we can always open a new issue if we see it again, thanks @yaronyg

**Comment by kartikagarwal1993** [comment]
Even I faced the same error with pouchdb v5.3.2. I am trying to sync the database with remote couch db (v 1.6.1). Although if I change the heartbeat to false in replication, the error is not thrown. I noticed the difference while solving the issue of heartbeat as mentioned in #4541 . But luckily it solved this problem also.

**Comment by nolanlawson** [comment]
@kartikagarwal1993 Can you upgrade to the latest version of PouchDB and see if you still see the error? A lot has changed between 5.3.2 and 6.0.7.

**Comment by kartikagarwal1993** [comment]
@nolanlawson Yes sure I can try that.

**Comment by kartikagarwal1993** [comment]
@nolanlawson I upgraded the version to 6.0.7. But the error is still thrown. The error creeps in after a certain amount of time only.

**Comment by PrasannaVenkadesh** [comment]
@nolanlawson I am also using 6.0.7. On Both firefox 50 and Chromium 54 the error peeps in. I have live sync enabled. Once this error pops in, no live sync happens. I have to refresh the page for the sync to happen and this error peeps in again after sometime.

![pouch_db_error](https://cloud.githubusercontent.com/assets/788244/21074812/304ebc22-bf28-11e6-852e-be1cdab21c9d.png)

**Comment by nolanlawson** [comment]
Hey folks, if anybody could provide a test case to reproduce, that would be extremely helpful. Looking at screenshots and console output, I am unable to figure out what's going on.

**Comment by PrasannaVenkadesh** [comment]
@nolanlawson I am not running anything complex. I am just trying out the [Getting Started Example](https://pouchdb.com/getting-started.html) given in the site with minor change. Instead of using in-browser storage, I am using **pouchdb-server** and `http://localhost:5984/todos` as the endpoint for local db variable.

I have another computer which resides on the LAN itself and I also run another **pouchdb-server** on it. Both are listening on `0.0.0.0` port 5984. I have replaced the `remoteCouch` on the getting started code with `http://<ip_of_another_machine>:5984` and the sync between them are set to be `live` according to the code and this pops up.

**Comment by hpgmiskin** [comment]
I know there has been a number of pasted console logs on this same issue however would like to include another in case it is useful. We are experiencing much the same issue as @PrasannaVenkadesh described but we are using PouchDB on a hybrid mobile app developed using Ionic.

![screen shot 2016-12-14 at 21 23 19](https://cloud.githubusercontent.com/assets/5551702/21201831/51872040-c244-11e6-8b35-8e07eec66ff0.jpg)

Is there any more information we can provide to assist in identification of the issue? Our remote CouchDB is hosted on Cloudant but we have had the same issues when developing with a local CouchDB database. When we first started getting the error we tailed the logs and found that no requests were being made to the local CouchDB database.

**Comment by htrampe** [comment]
Same issue here - is there a workaround? Using a couch-db-server...THX

**Comment by theutopianster** [comment]
Hey guys I had the same issue with running sync or replicate from puch enabled website to couch db.
I finally figured out how to fix it, at least on my side.
Basically the timeout parameter was the culprit because after I set the variable to 0 inside pouch js it worked.
So in the file puchdb-6.1.2.js method `api.allDocs` 
I set` opts.timeout = 0;` and commented  
`('timeout' in opts) ? opts.timeout :
       ('timeout' in ajaxOpts) ? ajaxOpts.timeout :
       30 * 1000;`
I tested with 30 min, 1 hour and more and it does not die anymore.
I ran this with CouchDB 1.6.1 and 2.0 so I hope this works for you as well.

**Comment by smithmajor** [comment]
Hi, i did it like theutopianster and it's true, CouchDb is running well. Configuration: 1.6.1 on Windows, http Port 5984, https 6984 over Apache Webserver as Proxy. Used for GPS App. Tested with 5 cellphones simultan, no problem. Thanks a lot

**Comment by ict-programmer** [comment]
I was receiving the same error, but it is fixed now. The error cause: not canceling all PouchDB syncs on user logout. The error got resolved by canceling all syncs on user logout, like this:

```
syncHandler=this.localDB.sync(this.remoteDB,{live:true, retry:true}).on('complete',res=>{
    console.log('Sync is canceled')
})

this.userDB.logout().then(res=>{
    this.syncHandler.cancel()
}).catch(err=>{console.log('err -> ', err)})
```

**Comment by gkiko** [comment]
I also experienced this problem with pouch 7.0.0. I closed all database connections with `syncHandler.cancel()` as shown above. When I wanted to resume sync it was hanging because browsers don't allow more than 6 parallel connections (I have 5 databases open at once).

I solved this problem by cancelling a sync with `cancel` and then calling `close` for all databases. After this I can call `sync` again for all my databases.
